tests/bluetooth/bsim_bt/bsim_test_l2cap_stress/src/main.c

```diff
@@ -0,0 +1,443 @@
+
+	int ret = bt_l2cap_chan_send(chan, buf);
+
+	if (ret < 0) {
+		FAIL("L2CAP error %d\n", ret);
+		net_buf_unref(buf);
+	}
+
+	LOG_DBG("sent %d len %d", ret, len);
+	return ret;
+}
+
+struct net_buf *alloc_seg_cb(struct bt_l2cap_chan *chan)
+{
+	struct net_buf *buf = net_buf_alloc(&segment_pool, K_NO_WAIT);
+
+	ASSERT(buf, "Ran out of segment buffers");
+
+	return buf;
+}
+
+struct net_buf *alloc_buf_cb(struct bt_l2cap_chan *chan)
+{
+	return net_buf_alloc(&sdu_rx_pool, K_NO_WAIT);
+}
+
+static int get_l2cap_chan(struct bt_l2cap_chan *chan)
+{
+	for (int i = 0; i < L2CAP_CHANS; i++) {
+		if (l2cap_chans[i] == chan) {
+			return i;
+		}
+	}
+
+	FAIL("Channel %p not found\n", chan);
+	return -1;
+}
+
+static void register_channel(struct bt_l2cap_chan *chan)
+{
+	int i = get_l2cap_chan(NULL);
+
+	l2cap_chans[i] = chan;
+}
+
+void sent_cb(struct bt_l2cap_chan *chan)
+{
+	LOG_DBG("%p", chan);
+	int idx = get_l2cap_chan(chan);
+
+	if (tx_left[idx]) {
+		tx_left[idx]--;
+		l2cap_chan_send(chan, tx_data, sizeof(tx_data));
+	} else {
+		LOG_DBG("Done sending %p", chan->conn);
+	}
+}
+
+int recv_cb(struct bt_l2cap_chan *chan, struct net_buf *buf)
+{
+	LOG_DBG("len %d", buf->len);
+	rx_cnt++;
+
+	return 0;
+}
+
+void l2cap_chan_connected_cb(struct bt_l2cap_chan *l2cap_chan)
+{
+	struct bt_l2cap_le_chan *chan =
+		CONTAINER_OF(l2cap_chan, struct bt_l2cap_le_chan, chan);
+
+	SET_FLAG(flag_l2cap_connected);
+	LOG_DBG("%x (tx mtu %d mps %d) (tx mtu %d mps %d)",
+		l2cap_chan,
+		chan->tx.mtu,
+		chan->tx.mps,
+		chan->rx.mtu,
+		chan->rx.mps);
+
+	register_channel(l2cap_chan);
+}
+
+void l2cap_chan_disconnected_cb(struct bt_l2cap_chan *chan)
+{
+	UNSET_FLAG(flag_l2cap_connected);
+	LOG_DBG("%x", chan);
+}
+
+static struct bt_l2cap_chan_ops ops = {
+	.connected = l2cap_chan_connected_cb,
+	.disconnected = l2cap_chan_disconnected_cb,
+	.alloc_buf = alloc_buf_cb,
+	.alloc_seg = alloc_seg_cb,
+	.recv = recv_cb,
+	.sent = sent_cb,
+};
+
+struct bt_l2cap_le_chan *get_free_l2cap_le_chan(void)
+{
+	for (int i = 0; i < L2CAP_CHANS; i++) {
+		struct bt_l2cap_le_chan *le_chan = &l2cap_channels[i];
+
+		if (le_chan->state != BT_L2CAP_DISCONNECTED) {
+			continue;
+		}
+
+		memset(le_chan, 0, sizeof(*le_chan));
+		return le_chan;
+	}
+
+	return NULL;
+}
+
+int server_accept_cb(struct bt_conn *conn, struct bt_l2cap_chan **chan)
+{
+	struct bt_l2cap_le_chan *le_chan = NULL;
+
+	le_chan = get_free_l2cap_le_chan();
+	if (le_chan == NULL) {
+		return -ENOMEM;
+	}
+
+	memset(le_chan, 0, sizeof(*le_chan));
+	le_chan->chan.ops = &ops;
+	le_chan->rx.mtu = SDU_LEN;
+	le_chan->rx.init_credits = INIT_CREDITS;
+	le_chan->tx.init_credits = INIT_CREDITS;
+	*chan = &le_chan->chan;
+
+	return 0;
+}
+
+static struct bt_l2cap_server test_l2cap_server = {
+	.accept = server_accept_cb
+};
+
+static int l2cap_server_register(bt_security_t sec_level)
+{
+	test_l2cap_server.psm = 0;
+	test_l2cap_server.sec_level = sec_level;
+
+	__ASSERT_NO_MSG(bt_l2cap_server_register(&test_l2cap_server) == 0);
```

Comment:
Nack from me as well. Doesn't really matter if you're doing this intentionally or not - asserts containing code with side-effects is just generally a bad (and potentially dangerous) habit, and we shouldn't have such code anywhere. You never know who's going to be looking at it and decide to copy it for some other purpose. Btw, `__ASSERT_EVAL()` might be convenient here (not 100% sure).